https://issues.redhat.com/browse/ACM-28571 (#8594)
open-cluster-management/rhacm-docs: troubleshooting/trouble_manifest_webhook.adoc

[#troubleshooting-manifest-webhook-limits]

> = Troubleshooting ManifestWork webhook limits

Reviewer: I am having a tough time with this title. The order of resources is MultiClusterRoleAssignment -> ClusterPermission -> ManifestWork. I think we should stick with MultiClusterRoleAssignment in the title, as it will be the direct top-level resource a customer is working with that is having this issue.

Troubleshooting MultiClusterRoleAssignment size limits

WDYT?

Author: Thanks for pointing that out! That seems more accurate to me too

> = Troubleshooting ManifestWork webhook limits

Author: ```suggestion
= Troubleshooting MultiClusterRoleAssignment size limits
```

> = Troubleshooting MultiClusterRoleAssignment size limits

Reviewer 2: ```suggestion
= Troubleshooting _MultiClusterRoleAssignment_ size limits
```

> When you enable fine-grained role-based access control for virtualzation, the `MultiClusterRoleAssignment` controller aggregates `roleAssignment` entries from `MultiClusterRoleAssignment` resources into a single shared `ClusterPermission` resource for each managed cluster namespace.

Reviewer 2: ```suggestion
When you enable fine-grained role-based access control for virtualization, the `MultiClusterRoleAssignment` controller aggregates `roleAssignment` entries from `MultiClusterRoleAssignment` resources into a single shared `ClusterPermission` resource for each managed cluster namespace.
```

> If your `roleAssignments` lists exceeds the `ManifestWork` webhook size limit of 512 KB, new assignments for the cluster fail to apply.

Reviewer: ```suggestion
If your `roleAssignments` lists exceeds the `ManifestWork` manifests size limit of 512 KB for a cluster, new assignments for the cluster fail to apply.
```

Reviewer: I think we should change all references of webhook to "manifests size limit". The webhook is just enforcing the limit, but the limit itself is related to the manifest size.

> If your `roleAssignments` lists exceeds the `ManifestWork` manifests size limit of 512 KB for a cluster, new assignments for the cluster fail to apply.

Reviewer 2: ```suggestion
If your `roleAssignments` list exceeds the `ManifestWork` manifests size limit of 512 KB for a cluster, new assignments for the cluster fail to apply.
```

> == Symptoms: Troubleshooting ManifestWork webhook limits

Reviewer: ```suggestion
== Symptoms: Troubleshooting ManifestWork manifest size limits
```

> The following symptoms indicate that you have exceeded the webhook limits:

Reviewer: ```suggestion
The following symptoms indicate that you have exceeded the manifest size limits:
```

> - When you apply a valid `MultiClusterRoleAssignment` resource, the cluster permissions are missing because the permissions are not created on the managed cluster.

Reviewer: ```suggestion
- When you apply a valid `MultiClusterRoleAssignment` resource, the permissions are not created on the managed cluster.
```

> - The following error message appears in the logs of the `ClusterPermission` resource: `denied the request: the size of manifests is X bytes which exceeds the 512000 limit.`

Reviewer: ```suggestion
- The following error message appears in the logs of the `ClusterPermission` controller: `denied the request: the size of manifests is X bytes which exceeds the 512000 limit.`
```

Reviewer: Logs are only in controller, not shown on resource.

> == Identifying the problem: Troubleshooting ManifestWork webhook limits

Reviewer: ```suggestion
== Identifying the problem: Troubleshooting ManifestWork manifest size limits
```

> To confirm if the `ManifestWork` webhook limit is the problem, check the `mra-managed-permissions` `ClusterPermission` resource. Complete the following steps:

Reviewer: ```suggestion
To confirm if the `ManifestWork` manifest size limit is the problem, check the `mra-managed-permissions` `ClusterPermission` resource. Complete the following steps:
```

. Compare the desired bindings with the active bindings. If the numbers do not match, the latest changes do not apply from the `mra-managed-permissions` `ClusterPermission` resource. Replace `<managed-cluster-namespace>` with the namespace of your managed cluster.

.. Run the following command to count the desired bindings:

+
[source,bash]
----
oc get clusterpermission mra-managed-permissions -n <managed-cluster-namespace> -o yaml | grep "owner/" | wc -l
----

.. Run the following command to count the active bindings that are successfully applied:

+
[source,bash]
----
oc get clusterpermission mra-managed-permissions -n <managed-cluster-namespace> -o yaml | grep "reason: AppliedManifestComplete" | wc -l
----

. Check the current capacity of the `ClusterPermission` specification. When there are more than 360,000 bytes, you are at or exceeding the limit. Run the following command:

+
[source,bash]
----
oc get clusterpermission mra-managed-permissions -n <managed-cluster-namespace> -o jsonpath='{.spec}' | wc -c
----

> == Resolving the problem: Troubleshooting ManifestWork webhook limits

Reviewer: ```suggestion
== Resolving the problem: Troubleshooting ManifestWork manifest size limits
```

To resolve the problem, remove assignments from some of the `MultiClusterRoleAssignment` objects. Complete the following steps:

> . Identify `roleAssignments` in your `MultiClusterRoleAssignment` resources that you can manage separately.

Reviewer: ```suggestion
. Identify `roleAssignments` in your `MultiClusterRoleAssignment` resources that you can manage separately for the affected cluster.
```

> . Remove the `roleAssignments` that you can manage separately from the `spec.roleAssignments` list in your `mra-managed-permissions` `ClusterPermission` resource.

Reviewer: ```suggestion
. Remove the `roleAssignments` that you can manage separately from the `spec.roleAssignments` list in your `MultiClusterRoleAssignment` resources.
```

> . Recreate the `roleAssignment` permissions as individual `ClusterPermission` resources.

Author: ```suggestion
. Re-create the `roleAssignment` permissions as individual `ClusterPermission` resources.
```

> . Re-create the `roleAssignment` permissions as individual `ClusterPermission` resources.

Reviewer: ```suggestion
. Re-create the `roleAssignment` permissions as separate `ClusterPermission` resources.
```
